django/contrib/gis/db/backends/mysql/schema.py

```diff
@@ -81,3 +95,36 @@ def create_spatial_indexes(self):
                     f"support them.",
                 )
         self.geometry_sql = []
+
+    def _should_have_spatial_index(self, field):
+        # MySQL doesn't support spatial indexes on NULL columns
+        return (
+            isinstance(field, GeometryField) and field.spatial_index and not field.null
+        )
```

Comment:
It's quite confusing that we have here three different approaches:
- `_should_have_spatial_index()` hook on MySQL,
- `isinstance(old_field, GeometryField) and old_field.spatial_index` check on Oracle, and
- `getattr(old_field, "spatial_index", False)` on PostgreSQL.

Can we add `_field_should_be_spatial_indexed(model, field)` hook on all backend? That would be a similar approach to the `BaseDatabaseSchemaEditor._field_should_be_indexed()`.

Do we need `isinstance(field, GeometryField)` check on MySQL and Oracle? Is it not enough to use `getattr(old_field, "spatial_index", False)` like on PostgreSQL?

Reply:
These backends all did it differently already and I did not want to risk a backwards incompatible change so I kept the checks in the exact same way. For the record, Spatialite also uses `isinstance(field, GeometryField) and field.spatial_index`.

That said, I agree that it's confusing and likely unnecessary. I think this stuff really belongs in a util function or mixin in `gis.db.backends.base` that all backends use. I would favor the `isinstance` check as it's more specific, but I would check for `BaseSpatialField` instead as any subclasses of it should support spatial indexes (including both `GeometryField` and `RasterField`). MySQL would then also check for `not field.null` as it's a special case in that regard.

Maybe we should not change this here and instead do a separate PR with this refactoring? Let me know what you think.

Reply:
We should split it into multiple cleanups before fixing ticket-35074.

For example, lack of `not field.null` in the `remove_field()` on Oracle seems to be a good candidate as a separate bugfix. Next, I'd introduce `_field_should_be_spatial_indexed()` and `_spatial_index_sql()` on Oracle, MySQL, and PostgreSQL without changing the current behavior. Finally, we can fix ticket-35074.